lib/internal/modules/esm/resolve.js

```diff
@@ -938,19 +948,77 @@ function defaultResolve(specifier, context = {}, defaultResolveUnused) {
   let parsed;
   try {
     parsed = new URL(specifier);
-    if (parsed.protocol === 'data:') {
+    if (!parentURL) {
+      // allow
+    }
+    if (parsed.protocol === 'data:' ||
+        (experimentalHttpsModules && (
+          parsed.protocol === 'https:' ||
+          parsed.protocol === 'http:'))) {
       return {
-        url: specifier
+        url: specifier,
       };
     }
-  } catch {}
-  if (parsed && parsed.protocol === 'node:')
-    return { url: specifier };
-  if (parsed && parsed.protocol !== 'file:' && parsed.protocol !== 'data:')
-    throw new ERR_UNSUPPORTED_ESM_URL_SCHEME(parsed);
+  } catch { }
+  if (parentURL) {
+    const parsedParentURL = new URL(parentURL);
+    if (parsedParentURL.protocol === 'http:' ||
+        parsedParentURL.protocol === 'https:') {
+      if (shouldBeTreatedAsRelativeOrAbsolutePath(specifier)) {
+        parsed = new URL(specifier, parsedParentURL);
+        // data: and blob: disallowed due to allowing file: access via
+        // indirection
+        if (parsed &&
+            parsed.protocol !== 'https:' &&
+            parsed.protocol !== 'http:') {
+          throw new ERR_UNSUPPORTED_ESM_NETWORK_IMPORT_SPECIFIER(
+            specifier,
+            parentURL,
+            'cannot import local location from remote location,' +
+            'load a it via a remote location.');
+        } else {
+          return { url: parsed.href };
+        }
+      } else {
+        throw new ERR_UNSUPPORTED_ESM_NETWORK_IMPORT_SPECIFIER(
+          specifier,
+          parentURL,
+          'only relative and absolute specifiers are supported.');
+      }
+    }
+  }
+  if (parsed && parsed.protocol === 'node:') return { url: specifier };
+  if (
+    parsed &&
+    parsed.protocol !== 'file:' &&
+    parsed.protocol !== 'data:' &&
+    (
+      !experimentalHttpsModules || (
+        parsed.protocol !== 'https:' &&
+        parsed.protocol !== 'http:')
+    )
+  )
+    throw new ERR_UNSUPPORTED_ESM_URL_SCHEME(parsed, [
+      'file',
+      'data',
+      ...(experimentalHttpsModules ? ['https', 'http'] : []),
+    ]);
```

Comment:
I would really like to see this whole block of code cleaned up a bit. Is there a higher level construct we can reflect via a function call?

This is the start of the main resolver - it should be crystal clear to read it, and I feel like these 40 lines are very hard to decipher.

Reply:
do you just mean can we extract this to its own function?

Reply:
Yes, as much as possible the top-level resolve function should be a clear set of high-level routines for maintainability, with actual implementations in the subroutines.

Reply:
done. I created a new `checkIfDisallowedImport()` to capture a bunch of these related throws; it's not as pure as I'd like, but I think sufficient for now.

I plan to do a housekeeping PR after this one is one to add code docs and tidy things, so I can polish some of this gnarliness then if desired.